server/automute.go

```diff
@@ -0,0 +1,182 @@
+package main
+
+import (
+	"database/sql"
+	"fmt"
+	"strconv"
+
+	"github.com/mattermost/mattermost/server/public/model"
+	"github.com/pkg/errors"
+)
+
+const (
+	NotifyPropAutomuted = "msteams_automute"
+
+	// This preference tracks if we've enabled automuting for the user by muting all of their connected channels.
+	PreferenceNameAutomuteEnabled = "automute_enabled"
+)
+
+// enableAutomute mutes all of the user's linked channels and DMs and sets a preference to track that automute is
+// enabled. It assumes that the caller has checked that the user is both connected and has their primary platform
+// set to MS Teams.
+func (p *Plugin) enableAutomute(userID string) (bool, error) {
+	return p.setAutomuteEnabledForUser(userID, true)
+}
+
+// disableAutomute mutes all of the user's linked channels and DMs and sets a preference to track that automute is
+// disabled. It assumes that the user is either not connected to MS Teams or has their primary platform set to MM.
+func (p *Plugin) disableAutomute(userID string) (bool, error) {
+	return p.setAutomuteEnabledForUser(userID, false)
+}
+
+func (p *Plugin) setAutomuteEnabledForUser(userID string, automuteEnabled bool) (bool, error) {
+	if channelsAutomuted := p.getAutomuteIsEnabledForUser(userID); channelsAutomuted == automuteEnabled {
+		// We've already automuted all the users' channels, so there's nothing to do
+		return false, nil
+	}
+
+	var membersToMute []*model.ChannelMemberIdentifier
+
+	channels, appErr := p.API.GetChannelsForTeamForUser("", userID, true)
+	if appErr != nil {
+		return false, errors.Wrap(appErr, fmt.Sprintf("Unable to get channels for user %s to automute them", userID))
+	}
+
+	for _, channel := range channels {
+		if linked, err := p.canAutomuteChannel(channel); err != nil {
+			return false, err
+		} else if !linked {
+			continue
+		}
+
+		membersToMute = append(membersToMute, &model.ChannelMemberIdentifier{ChannelId: channel.Id, UserId: userID})
+	}
+
+	if err := p.setChannelMembersAutomuted(membersToMute, automuteEnabled); err != nil {
+		return false, errors.Wrap(err, fmt.Sprintf("Unable to automute all channels for linked user %s", userID))
+	}
+
+	if err := p.setAutomuteIsEnabledForUser(userID, automuteEnabled); err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+func (p *Plugin) setChannelMembersAutomuted(members []*model.ChannelMemberIdentifier, automuteEnabled bool) error {
+	notifyProps := getNotifyPropsForAutomute(automuteEnabled)
+
+	i := 0
+	perPage := 200
+	for i < len(members) {
```

Comment:
Should we inline this?
```go
for i := 0; i < len(members); i += perPage {
    // ...
}
```

Reply:
For some reason, that looks weird to me, but I can't think of a good reason not to do that. Maybe it's because it's rare we do a regular for loop in Go without it involving a `range`.

I'll change this regardless